central/complianceoperator/v2/report/manager/watcher/scanconfigwatcher.go

```diff
@@ -0,0 +1,196 @@
+		return nil
+	}
+}
+
+// Subscribe a new readyQueue to the watcher
+func (w *scanConfigWatcherImpl) Subscribe(id string) {
+	concurrency.WithLock(&w.snapshotsLock, func() {
+		w.scanConfigResults.ReportSnapshotIDs = append(w.scanConfigResults.ReportSnapshotIDs, id)
+	})
+}
+
+// Stop the watcher
+func (w *scanConfigWatcherImpl) Stop() {
+	w.cancel()
+}
+
+// Finished indicates whether the watcher is finished or not
+func (w *scanConfigWatcherImpl) Finished() concurrency.ReadOnlySignal {
+	return w.stopped
+}
+
+func (w *scanConfigWatcherImpl) run(timerC <-chan time.Time) {
+	defer func() {
+		w.stopped.Signal()
+		<-w.stopped.Done()
+	}()
+	for {
+		select {
+		case <-w.ctx.Done():
+			log.Infof("Stopping scan config watcher")
+			return
+		case <-timerC:
+			log.Warnf("Timeout waiting for the ScanConfiguration %s's scans to finish", w.scanConfigResults.ScanConfig.GetScanConfigName())
+			concurrency.WithLock(&w.snapshotsLock, func() {
+				w.scanConfigResults.Error = ScanConfigTimeoutError
+				w.readyQueue.Push(w.scanConfigResults)
+			})
+			return
+		case result := <-w.scanC:
+			if err := w.handleScanResults(result); err != nil {
+				log.Errorf("Unable to handle scan results %s: %v", result.Scan.GetScanName(), err)
+			}
+		}
+		if w.totalResults != 0 && w.totalResults == len(w.scanConfigResults.ScanResults) {
+			concurrency.WithLock(&w.snapshotsLock, func() {
+				w.readyQueue.Push(w.scanConfigResults)
+			})
+			return
+		}
+	}
+}
+
+func (w *scanConfigWatcherImpl) handleScanResults(result *ScanWatcherResults) error {
+	// Here we have the scan config id and the scan
+	if w.totalResults == 0 {
+		var profileNames []string
+		for _, p := range w.scanConfigResults.ScanConfig.GetProfiles() {
+			profileNames = append(profileNames, p.GetProfileName())
+		}
+		log.Debugf("Profiles %v", profileNames)
+		var clusters []string
+		for _, c := range w.scanConfigResults.ScanConfig.GetClusters() {
+			clusters = append(clusters, c.GetClusterId())
+		}
+		profileQuery := search.NewQueryBuilder().
+			AddExactMatches(
+				search.ComplianceOperatorProfileName,
+				profileNames...,
+			).
+			AddExactMatches(
+				search.ClusterID,
+				clusters...,
+			).ProtoQuery()
+		profiles, err := w.profileDS.SearchProfiles(result.Ctx, profileQuery)
+		if err != nil {
+			return err
+		}
+		for _, p := range profiles {
+			scanRefQuery := search.NewQueryBuilder().AddExactMatches(
+				search.ComplianceOperatorProfileRef,
+				p.GetProfileRefId(),
+			).ProtoQuery()
+			scans, err := w.scanDS.SearchScans(result.Ctx, scanRefQuery)
+			if err != nil {
+				log.Errorf("Unable to retrieve scans: %v", err)
+				continue
+			}
+			for _, s := range scans {
+				log.Debugf("Adding scan to wait %s", s.GetScanName())
+				w.scansToWait.Add(fmt.Sprintf("%s:%s", s.GetClusterId(), s.GetId()))
+			}
+		}
+		w.totalResults = len(w.scansToWait)
+		log.Infof("Scan config %s needs to wait for %d scans", w.scanConfigResults.ScanConfig.GetScanConfigName(), w.totalResults)
+	}
```

Comment:
This looks like a good candidate to be exported to a separate function.

Reply:
Will do! 👍

Reply:
Done